tokio/src/fs/file.rs

```diff
@@ -753,20 +808,75 @@ impl AsyncWrite for File {
                     let n = buf.copy_from(src, me.max_buf_size);
                     let std = me.std.clone();
 
-                    let blocking_task_join_handle = spawn_mandatory_blocking(move || {
-                        let res = if let Some(seek) = seek {
-                            (&*std).seek(seek).and_then(|_| buf.write_to(&mut &*std))
-                        } else {
-                            buf.write_to(&mut &*std)
-                        };
+                    #[allow(unused_mut)]
+                    let mut data = Some((std, buf));
+
+                    let mut task_join_handle = None;
+
+                    #[cfg(all(
+                        tokio_unstable,
+                        feature = "io-uring",
+                        feature = "rt",
+                        feature = "fs",
+                        target_os = "linux"
+                    ))]
+                    {
+                        use crate::runtime::Handle;
+
+                        // Handle not present in some tests?
+                        if let Ok(handle) = Handle::try_current() {
+                            if handle.inner.driver().io().check_and_init()? {
+                                task_join_handle = {
+                                    use crate::runtime::driver::op::Op;
+
+                                    let (std, buf) = data.take().unwrap();
+                                    if let Some(seek) = seek {
+                                        // we do std seek before a write, so we can always use u64::MAX (current cursor) for the file offset
+                                        // seeking only modifies kernel metadata and does not block, so we can do it here
+                                        (&*std).seek(seek).map_err(|e| {
+                                            io::Error::new(
+                                                e.kind(),
+                                                format!("failed to seek before write: {e}"),
+                                            )
+                                        })?;
+                                    }
+
+                                    let op = Op::write_at(std, buf, u64::MAX);
+
+                                    let handle = BoxedOp(Box::pin(async move {
+                                        let (r, buf, _fd) = op.await;
+                                        match r {
+                                            Ok(_n) => (Operation::Write(Ok(())), buf),
```

Comment:
I think there is a chance that the buffer is not fully drained here.
`Write(Ok())` should be returned only if `buf.is_empty()`.

I think you need to wrap the write in a loop and handle it with something like:
```rust
Ok(0) => break (Operation::Write(Err(io::ErrorKind::WriteZero.into())), buf),
Ok(_) if buf.is_empty() => break (Operation::Write(Ok(())), buf),
Ok(_) => continue, // more to write
Err(e) => break (Operation::Write(Err(e)), buf),
```